### ISSUE TITLE
local options do not work with install on local packages (was: Run okay with profiling enabled, but installed exe has NO profiling enabled)

### ISSUE THREAD

**Issue by complyue** [haskell/cabal#7297]
**Describe the bug**
Can't install profiling-enabled exe with cabal 3.2.0.0, while direct running works correctly.

**To Reproduce**

```
$ cabal v2-run --enable-profiling --enable-executable-profiling  exe:els -- +RTS --info                                  
Resolving dependencies...
Up to date
 [("GHC RTS", "YES")
 ,("GHC version", "8.8.3")
 ,("RTS way", "rts_thr_p")
 ,("Build platform", "x86_64-apple-darwin")
 ,("Build architecture", "x86_64")
 ,("Build OS", "darwin")
 ,("Build vendor", "apple")
 ,("Host platform", "x86_64-apple-darwin")
 ,("Host architecture", "x86_64")
 ,("Host OS", "darwin")
 ,("Host vendor", "apple")
 ,("Target platform", "x86_64-apple-darwin")
 ,("Target architecture", "x86_64")
 ,("Target OS", "darwin")
 ,("Target vendor", "apple")
 ,("Word size", "64")
 ,("Compiler unregisterised", "NO")
 ,("Tables next to code", "YES")
 ,("Flag -with-rtsopts", "-maxN31 -qg")
 ]
$ cabal v2-install --enable-profiling --overwrite-policy=always --enable-executable-profiling exe:els && els +RTS --info 
Wrote tarball sdist to
 ...
Resolving dependencies...
Up to date
Copying 'els'
 [("GHC RTS", "YES")
 ,("GHC version", "8.8.3")
 ,("RTS way", "rts_thr")
 ,("Build platform", "x86_64-apple-darwin")
 ,("Build architecture", "x86_64")
 ,("Build OS", "darwin")
 ,("Build vendor", "apple")
 ,("Host platform", "x86_64-apple-darwin")
 ,("Host architecture", "x86_64")
 ,("Host OS", "darwin")
 ,("Host vendor", "apple")
 ,("Target platform", "x86_64-apple-darwin")
 ,("Target architecture", "x86_64")
 ,("Target OS", "darwin")
 ,("Target vendor", "apple")
 ,("Word size", "64")
 ,("Compiler unregisterised", "NO")
 ,("Tables next to code", "YES")
 ,("Flag -with-rtsopts", "-maxN31 -qg")
 ]
$ els +RTS -p
els: the flag -p requires the program to be built with -prof
els: 
els: Usage: <prog> <args> [+RTS <rtsopts> | -RTS <args>] ... --RTS <args>
els: ...
$ 
```

**Expected behavior**
The built executable can have `+RTS -p` enabled

**System information**
 - Operating system

macOS Mojave

 - `cabal`, `ghc` versions

```
$ cabal --version
cabal-install version 3.2.0.0
compiled using version 3.2.0.0 of the Cabal library 
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.8.3
```

**Additional context**

**Comment by fendor** [comment]
Looks like a duplicate of #5982

**Comment by complyue** [comment]
In addition to the cmdl options being ignored (as with #5982), I also instrumented in the `cabal.project.local` for profiling, hope this add more information (or concerns) for the diagnostics.

```
$ cat cabal.project.local 

profiling: True
profiling-detail: all-functions

optimization: 0

$ cat cabal.project

-- Please change Cabal project settings per your needs

-- FOLLOWING CONTENTS GONNA BE OVERWRITTEN BY EPM --

--  * this file should be:
--  /fw/m3cyue/edh-universe/cabal.project

packages:
 ./e-wrks/edh/lossless-decimal/lossless-decimal.cabal
 ./e-wrks/edh/elr/elr.cabal
 ./e-wrks/edh/host.hs/edh.cabal
 ./e-wrks/nedh/host.hs/nedh.cabal
 ./e-wrks/sedh/host.hs/sedh.cabal
 ./e-wrks/hasdim/host.hs/hasdim.cabal
 ./e-wrks/haskit/host.hs/haskit.cabal
 ./e-wrks/els/host.hs/els.cabal
$ 
```

**Comment by complyue** [comment]
BTW, Albert Y.C. Lai suggests in a Cafe list mail that such a workaround should work, but seemingly not for my case.

 https://mail.haskell.org/pipermail/haskell-cafe/2021-February/133453.html

**Comment by amigalemming** [comment]
Do you have profiling disabled globally in `.cabal/config`?

**Comment by complyue** [comment]
@amigalemming I'm not sure, here is the content:

```
$ cat .cabal/config | grep -e '^\s*--' -v


repository mirrors.tuna.tsinghua.edu.cn
  url: http://mirrors.tuna.tsinghua.edu.cn/hackage




remote-repo-cache: /Users/cyue/.cabal/packages
world-file: /Users/cyue/.cabal/world
extra-prog-path: /Users/cyue/.cabal/bin
build-summary: /Users/cyue/.cabal/logs/build.log
remote-build-reporting: anonymous
jobs: $ncpus
install-method: copy
installdir: /Users/cyue/.cabal/bin

haddock

init

install-dirs user

install-dirs global

program-locations

program-default-options
$
```

Did I?

**Comment by amigalemming** [comment]
On Thu, 4 Mar 2021, Compl Yue wrote:

> @amigalemming I'm not sure, here is the content:
> 
> $ cat .cabal/config | grep -e '^\s*--' -v

I would grep for 'profil', but it looks like you have not touched the 
profiling settings.

**Comment by romanofski** [comment]
I actually tripped over this too when working on  purebred-mua/purebred#299. Also not explicitly profiling disabled. What I thought was interesting was that setting `profiling: True` in my `cabal.project.local` would not have an effect and neither would `--enable-library-profiling`. And by affect I mean, that under `dist-newstyle` I find profiling information, but in the installed package under `~/.cabal/store` just the normal libraries.

Even the build progress does not show that `profiling` is enabled. If I build with:

```
$ cabal new-install -p -j --overwrite-policy=always --lib purebred-icu
```
I get:
```
In order, the following will be built (use -v for more details):
 - purebred-0.1.0.0 (lib) (requires build)
 - purebred-icu-0.1.0.0 (lib) (requires build)
```
Using the workaround linked from the mailing list, I get:
```
In order, the following will be built (use -v for more details):
 - purebred-0.1.0.0 (lib)  --enable-profiling (requires build)
 - purebred-icu-0.1.0.0 (lib)  --enable-profiling (requires build)
```
Perhaps it has something to do with multiple packages that even the command line arguments are not applied? My config that enabled profiling:
```
packages: .
          ../purebred-icu

package *
  profiling: True
```
The one which does not enable does not have the `package *` line.

**Comment by Mikolaj** [comment]
That's a long shot, but it may be related to #7236, where static compilation works with `cabal build`, but is ignored with `cabal install`.

**Comment by fgaz** [comment]
I did a few tests and it looks like my hypothesis is correct.

---

### Workaround

for every local package `x`, add this snippet to your cabal.project

```cabal.project
package `x`
  profiling: True
```

The same applies for other flags.

---

I'll edit the title to reflect this.
This is very likely to be a part of what's happening in #7236

**Comment by jneira** [comment]
I am starting to think that actual implementation of cabal install is giving us more problems than a raw build + copy bins would do. Another symptom: the popularization of the pattern `cabal build && copy $(cabal list-bins myexec) somepath`

Afaik the actual mechanism was implemented to make the code reuse the `cabal install my.local.or.remote.tar` variation. am I correct?

**Comment by gbaz** [comment]
Kudos to @fgaz for the excellent sleuthing here.

I think this is really a bad interaction of how install treats cli flags -- it should probably translate them to package local flags for the package being installed, rather than top level project flags.

**Comment by Mikolaj** [comment]
I think users expect flags to do the same for `cabal build` and `cabal install`, even just to be able to easily switch between the two in commandline invocations. Does your proposal ensure that for each use case of `cabal install`? If not, perhaps that's the criterion for which of many functionalities of `cabal install` should be split off and renamed to a less generic verb so that it's clear the semantics is very different from `cabal build`.

Actually, perhaps users even expect `cabal install` to be equilvalent to what @jneira cited: `cabal build && cp $(cabal list-bins myexec) somepath`.

**Comment by fgaz** [comment]
By the way this is also super counterintuive wrt explicitly named remote targets. For example `cabal install --enable-profiling some-package-from-hackage` will _not_ enable profiling for that package!

**Comment by jneira** [comment]
to let it clear, the welcomed pr would follow @gbaz proposal?:

> I think this is really a bad interaction of how install treats cli flags -- it should probably translate them to package local flags for the package being installed, rather than top level project flags.
 
Either for installing within a local project or from hackage/tarball?

**Comment by jneira** [comment]
Question: top level cabal.project config, only applied to local packages, should be applied to any package being installed, like cli flags?

It is clear that `cabal install package-from-hackage-or-tarball --enable-profiling` should build the package with profiling enabled.
Not doing so breaks clearly user expectations.

However if i have in my cabal.project

```
profiling: true
```

should be applied when doing a  `cabal install package-from-hackage-or-tarball`, or it would be somewhat surprising?

**Comment by gbaz** [comment]
Is the top level project consulted at all when doing an install? My recollection is "no".

**Comment by jneira** [comment]
hmm not sure what is the meaning of consulted here, it is not being applied (or it would be a workaround for the issue) but it is available in the config for build prior to the install  (like the `package` entries in the cabal.project, the actual workaround). 

So we could use it like the cli flags (which also are available and are not applied)

**Comment by jneira** [comment]
If users dont want to apply them they always could do a `cabal install package-from-hackage-or-tarball --ignore-project`

**Comment by gbaz** [comment]
I'm honestly not sure of the right behavior here and could be swayed either way. Would your proposal amount to effectively treating the sdist'd tarball being installed as a local package?

**Comment by Mikolaj** [comment]
Slightly related: we really should split `cabal install-local` and `cabal install-from-hackage-or-tarball` both in UI and as much as feasable in the code paths.

**Comment by jneira** [comment]
Well, to be fair i *personally* would expect cabal.project top level build config be applicable only installing local packages executables and no remote ones, even if they are included in dependencies. It is clear to me than top project config only is applicable to local packages and i think it is described in docs. But i am not totally convinced so asked to know other's opinion.

So to resuming in a table: 

|| build local pkg | install local pkg  | install remote pkg |
|-|---------------|------------------|--------------------|
| top project config | ✔️| ✔️ | ❌|
| cli args | ✔️| ✔️|✔️|

All would be a little bit clearer (and less convenient) if top level project config did not exist and you had to do `package :locals\n xxx`

**Comment by jneira** [comment]
> I'm honestly not sure of the right behavior here and could be swayed either way. Would your proposal amount to effectively treating the sdist'd tarball being installed as a local package?

yeah so adding the package(s) as a local package(s) in the project config could do the fix for all target types

I foresee that would be simpler than distinguish cli flags (they are not marked as such in the actual project config afaics) from project config top level ones and apply them to remote packages. In that case we still could add local packages targets as such in the project config

**Comment by jneira** [comment]
[This issue](https://github.com/haskell/cabal/issues/2997) is about `--extra-lib-dirs` and `--extra-include-dirs` being ignored as cli args, in cabal.project inside package sections but is has interesting insights (from hvr, phadej, emilpy and others) about they being ignored as cli args for `cabal install` specifically

**Comment by alt-romes** [comment]
I've observed some quite interesting behaviour regarding this and related issues (#7236, #8909), which report that flags passed to install are ignored and not applied to the executable being built.

Namely, note the following
```
cabal install --enable-profiling # NOT profiled
cabal install <my-pkg> --enable-profiling # DOES profile
cabal install exe:<my-pkg> --enable-profiling # NOT profile
```

I'm investigating...

**Comment by alt-romes** [comment]
Fix in #9697.

**Comment by alt-romes** [comment]
To add a bit to the discussion:
At a certain point (76670ebddeb0a914d816f4420c511c1bc1f044eb) we started applying *local* configuration (such as cli arguments and top level cabal.project fields) to the packages being installed. However, this patch only accounted for exact package names being specified as `cabal install` targets.

So
```
cabal install <my-pkg> --enable-profiling
```
would apply `--enable-profiling` to the package, but any other form of target like `all:exes`, no explicit target, `exe:my-exe`, ... would not get the local configuration.
